src/java.base/share/classes/java/util/jar/JarFile.java

```diff
@@ -597,25 +598,14 @@ private String getBasename(String name) {
     }
 
     private JarEntry getVersionedEntry(String name, JarEntry defaultEntry) {
-        if (!name.startsWith(META_INF)) {
-            int[] versions = JUZFA.getMetaInfVersions(this, name);
-            if (BASE_VERSION_FEATURE < versionFeature && versions.length > 0) {
-                // search for versioned entry
-                for (int i = versions.length - 1; i >= 0; i--) {
-                    int version = versions[i];
-                    // skip versions above versionFeature
-                    if (version > versionFeature) {
-                        continue;
-                    }
-                    // skip versions below base version
-                    if (version < BASE_VERSION_FEATURE) {
-                        break;
-                    }
-                    JarFileEntry vje = (JarFileEntry)super.getEntry(
-                            META_INF_VERSIONS + version + "/" + name);
-                    if (vje != null) {
-                        return vje.withBasename(name);
-                    }
+        if (BASE_VERSION_FEATURE < versionFeature && !name.startsWith(META_INF)) {
+            BitSet versions = JUZFA.getMetaInfVersions(this, name);
+            int version = versions.nextSetBit(BASE_VERSION_FEATURE);
+            while (version >= BASE_VERSION_FEATURE && version <= versionFeature) {
```

Comment:
The while loop never updates `version`, so the body will only ever run once, right?

Also, this changes the search order for versions, the previous look searched backwards, from newest to oldest version?

Reply:
Yes, fixed traversal and we should search backwards from featureVersion. This actually simplifies the loop a bit, but I'll re-run tests.

(Also suggests we might be lacking some tests here, since no tests got stuck on that logic error :D)

Reply:
Good. I had only sanity tested java/util/zip locally but figures this higher-level stuff is tested via java/util/jar tests.